test/test_images/probe_helper/probe_helper.go

```diff
@@ -300,24 +349,65 @@ func receiveEvent(ctx context.Context, receivedEvents *receivedEventsMap) cloudE
 	}
 }
 
-func healthChecker(w nethttp.ResponseWriter, r *nethttp.Request) {
-	lastProbeEventTimestamp.RLock()
-	defer lastProbeEventTimestamp.RUnlock()
-	lastReceiverEventTimestamp.RLock()
-	defer lastReceiverEventTimestamp.RUnlock()
-	now := time.Now()
-	if (now.Sub(lastProbeEventTimestamp.time) > maxStaleTime) ||
-		(now.Sub(lastReceiverEventTimestamp.time) > maxStaleTime) {
-		w.WriteHeader(nethttp.StatusServiceUnavailable)
-		return
+// Options is the struct type containing the optional parameters to be passed to
+// the probe helper.
+type Options struct {
+	receiverListener net.Listener
+	probeListener    net.Listener
+	storageClient    *storage.Client
+	pubsubClient     *pubsub.Client
+}
+
+// Option is an option for the probe helper.
+type Option func(*Options)
+
+// WithStorageClient forces the probe helper to use the specified storage client
+// instead of creating its own.
+func WithStorageClient(c *storage.Client) Option {
+	return func(o *Options) {
+		o.storageClient = c
+	}
+}
+
+// WithPubSubClient forces the probe helper to use the specified pubsub client
+// instead of creating its own.
+func WithPubSubClient(c *pubsub.Client) Option {
+	return func(o *Options) {
+		o.pubsubClient = c
+	}
+}
+
+// WithReceiverListener forces the probe helper receiver to use the specified
+// listener instead of relying on the CloudEvents client to create its own.
+func WithReceiverListener(l net.Listener) Option {
```

Comment:
I was running into issues in the unit tests where the free ports would not in fact be free by the time they would need to be opened again, probably caused by some race, which is why I opted for using listeners directly. Seems fixed now without using listeners.

Reply:
I am not sure if using a listener will remove the flakiness completely. Since the server is started in a goroutine. Maybe worth some validation. But a retry also works and is probably more reliable.